src/PKSim.Infrastructure/Serialization/Json/SnapshotSerializer.cs

```diff
@@ -30,18 +38,39 @@ public class SnapshotSerializer : ISnapshotSerializer
             json = await reader.ReadToEndAsync();
          }
 
+         var schema = validateSnapshot(snapshotType);
          var deserializedSnapshot = JsonConvert.DeserializeObject(json, _settings);
-         //TODO. We should validate that the json actually matches our type. If not, we could throw an explicit error
-         //This should be done using  /jsonschema
+
          switch (deserializedSnapshot)
          {
             case JObject jsonObject:
-               return new[] {jsonObject.ToObject(snapshotType)};
+               return new[] {validatedObject(jsonObject, schema, snapshotType)};
+
             case JArray array:
-               return array.Select(x => x.ToObject(snapshotType)).ToArray();
+               return array.Select(x => validatedObject(x, schema, snapshotType)).ToArray();
             default:
                return null;
          }
       }
+
+      private object validatedObject(JToken jToken, JSchema schema, Type snapshotType)
+      {
+         if (jToken.IsValid(schema, out IList<string> errorMessages))
+            return jToken.ToObject(snapshotType);
+
+         throw new SnapshotFileMismatchException(snapshotType.Name, errorMessages);
+      }
+
+      private JSchema validateSnapshot(Type snapshotType)
+      {
+         return _schemas.GetOrAdd(snapshotType, createSchemaForType);
```

Comment:
cache the schema to avoid the limitation of 10 generation per hour with the free version. We do not need to generate it every time